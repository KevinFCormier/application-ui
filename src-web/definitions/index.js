/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import lodash from 'lodash'

import { RESOURCE_TYPES } from '../../lib/shared/constants'

import hcmpods from './hcm-pods'
import hcmnodes from './hcm-nodes'
import hcmpvs from './hcm-pvs'

const resourceData = {
  [RESOURCE_TYPES.HCM_PODS.name]: hcmpods,
  [RESOURCE_TYPES.HCM_NODES.name]: hcmnodes,
  [RESOURCE_TYPES.HCM_PVS.name]: hcmpvs
}

function getResourceData(resourceType) {
  var def = resourceData[resourceType.name]
  if (!def)
    //eslint-disable-next-line no-console
    console.error(`No resource data found for '${resourceType}'`)
  return def
}

export default getResourceData

export function getPrimaryKey(resourceType) {
  let def = getResourceData(resourceType)
  let pk = def.primaryKey

  if (!pk)
    pk = 'metadata.uid'

  return pk
}

export function getURIKey(resourceType) {
  let def = getResourceData(resourceType)
  let uriKey = def.uriKey

  if (!uriKey)
    uriKey = 'metadata.name'

  return uriKey
}

export function getDefaultSearchField(resourceType) {
  var def = getResourceData(resourceType)
  let sf = def && def.defaultSearchField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf)
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  return sf
}

export function getDefaultSortField(resourceType) {
  var def = getResourceData(resourceType)
  let sf = def && def.defaultSortField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf) {
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  }
  if (!sf)
    //eslint-disable-next-line no-console
    console.error(`No sortable fields defined for '${resourceType}' resource definition`)
  return sf
}

export function getLink(link, resource) {
  if(typeof(link) === 'boolean') {
    return `/${resource.metadata.namespace}/${resource.metadata.name}`
  } else if (typeof(link) === 'function') {
    return link(resource)
  } else {
    const parts = link.split('/')
    let path = ''
    parts.forEach(part => {
      const value = lodash.get(resource, part)
      path += `/${encodeURIComponent(value)}`
    })
    return path
  }
}
