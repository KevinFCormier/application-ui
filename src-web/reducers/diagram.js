/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import * as Actions from '../actions'
import { getFilterState, saveFilterState } from '../../lib/client/filter-helper'

const initialState = {
  diagramFilters: [],
}

export const diagram = (state=initialState, action) => {
  switch (action.type){
  default:
  case '@@INIT':{
    return initialState
  }
  case Actions.DIAGRAM_RESTORE_FILTERS: {
    const { namespace, name, initialDiagramFilters } = action
    const {filters: diagramFilters} = getFilterState(initialDiagramFilters, namespace, name)
    state.diagramFilters = diagramFilters
    return {...state}
  }
  case Actions.DIAGRAM_SAVE_FILTERS: {
    const { namespace, name, diagramFilters } = action
    saveFilterState(namespace, name, {filters:diagramFilters})
    state.diagramFilters = diagramFilters
    return {...state}
  }
  }
}
