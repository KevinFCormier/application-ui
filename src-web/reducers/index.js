/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/*
 *
 * Things you should never do inside a reducer:
 *
 * - Mutate its arguments
 * - Perform side effects like API calls and routing transitions
 * - Call non-pure functions, e.g. Date.now() or Math.random()
 *
 * Reducers must be deterministic pure functions.  Given the same arguments, it should calculate the next state and return it.
 * No surprises. No side effects. No API calls. No mutations. Just a calculation.
 *
 * Selectors should sit along side reducers.
 */

import { createResourceReducer, resourceReducerFunction } from './common'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import lodash from 'lodash'

export { user, loggedIn } from './user'
export { secondaryHeader } from './common'

export { modal } from './modal'

//the exported function name must match the resourceType value
export const HCMApplicationList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_APPLICATIONS))
export const HCMClusterList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_CLUSTERS))
export const HCMComplianceList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_COMPLIANCES))
export const HCMDashboardList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_DASHBOARD))
export const HCMNodeList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_NODES))
export const HCMPersistentVolumeList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PVS))
export const HCMPersistentVolumeClaimList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PVS_CLAIM))
export const HCMPodList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_PODS))
export const HCMPolicyList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_POLICIES))
export const HCMReleaseList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_RELEASES))
export const HCMRepositoryList = createResourceReducer(resourceReducerFunction, predicate.bind(null, RESOURCE_TYPES.HCM_REPOSITORIES))

export { default as catalog } from './catalog'
export { topology } from './topology'
export { resourceFilters } from './filter'

function predicate(resourceType, action) {
  if (lodash.isEqual(resourceType, action.resourceType))
    return true
  return lodash.find(lodash.values(resourceType), type => {
    if (typeof type === 'string') {
      return type.indexOf(action.resourceType) > -1
    }
    return false
  })
}
