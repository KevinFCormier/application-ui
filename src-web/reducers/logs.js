/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import {
  LOGS_RECEIVE_SUCCESS,
  LOGS_RECEIVE_FAILURE,
  LOGS_RECEIVE_IN_PROGRESS,
  LOGS_RESET,
  REQUEST_STATUS
} from '../actions'

export const logs = (state = null, action) => {
  switch (action.type) {
  case LOGS_RESET:
    return { logs: null, status: REQUEST_STATUS.DONE }
  case LOGS_RECEIVE_IN_PROGRESS:
    return { ...state, status: LOGS_RECEIVE_IN_PROGRESS }
  case LOGS_RECEIVE_SUCCESS:
    return { ...action, status: REQUEST_STATUS.DONE }
  case LOGS_RECEIVE_FAILURE:
    return {
      ...state,
      status: REQUEST_STATUS.ERROR,
      errorMessage: action.err && action.err.details,
      statusCode: action.err && action.err.statusCode
    }
  default:
    return state
  }
}
