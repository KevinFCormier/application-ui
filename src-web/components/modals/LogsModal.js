/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import ScrollBox from './ScrollBox'
import apolloClient from '../../../lib/client/apollo-client'
import { UPDATE_ACTION_MODAL } from '../../apollo-client/queries/StateQueries'
import { Modal, DropdownV2, Loading, Notification } from 'carbon-components-react'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import config from '../../../lib/shared/config'

resources(() => {
  require('../../../scss/logs.scss')
})

class LogsModal extends React.PureComponent {
  constructor(props) {
    super(props)
    this.client = apolloClient.getClient()
    this.state = {
      xhrPoll: false,
      loading: true,
      errors: ''
    }
  }

  componentWillMount() {
    if (parseInt(config['featureFlags:liveUpdates']) === 2) {
      const intervalId = setInterval(this.reload.bind(this), config['featureFlags:liveUpdatesPollInterval'])
      this.setState({ intervalId: intervalId })
    }
    if (this.props.data.item !== '') {
      const { data: { item } } = this.props
      this.setState({
        loading: false,
        selectedContainer: item.containers[0].name,
        containers: item.containers,
        containerName: item.containers[0].name,
        podName: item.metadata.name,
        podNamespace: item.metadata.namespace,
        clusterName: item.cluster.metadata.name
      })
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.data.item !== '') {
      const { data: { item } } = nextProps
      this.setState({
        selectedContainer: item.containers[0].name,
        containers: item.containers,
        containerName: item.containers[0].name,
        podName: item.metadata.name,
        podNamespace: item.metadata.namespace,
        clusterName: item.cluster.metadata.name,
        loading: false
      })
    }
    if (nextProps.data.errors !== '') {
      this.setState({
        errors: nextProps.data.errors
      })
    }
  }

  fetchLogs(containerName, podName, podNamespace, clusterName) {
    return apolloClient.getLogs(containerName, podName, podNamespace, clusterName).then(result => {
      if (result.data.logs.errors && result.data.logs.errors.length > 0){
        return result.data.logs.errors[0]
      } else {
        this.setState({
          logs: result.data.logs,
          loading: result.loading
        })
      }
    })
  }

  componentDidMount() {
    const { clusterName, containerName, loading, podName, podNamespace } = this.state
    if (!loading) {
      this.fetchLogs(containerName, podName, podNamespace, clusterName)
    }
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalId)
  }

  handleClose() {
    const { type } = this.props
    this.client.mutate({
      mutation: UPDATE_ACTION_MODAL,
      variables: {
        __typename: 'actionModal',
        open: false,
        type: type,
        resourceType: {
          __typename: 'resourceType',
          name: '',
          list: ''
        },
        data: {
          __typename:'ModalData',
          item: '',
          errors: ''
        }
      }
    })
  }

  render() {
    const { open } = this.props,
          { containers, loading, logs, podName, xhrPoll, errors } = this.state,
          { locale } = this.context

    return (
      <Modal
        id='view-logs-modal'
        className='logs'
        open={open}
        passiveModal
        modalLabel='Pod'
        modalHeading={podName}
        primaryButtonDisabled
        onRequestClose={this.handleClose.bind(this)}
        role='region'
        aria-label='logs'>
        <div className='logs-container'>
          {(errors !== '' && errors !== undefined)
            ? <Notification
              kind='error'
              title=''
              subtitle={errors} />
            : null}
          <div className='logs-container__actions'>
            <div className='dropdown-container'>
              <DropdownV2
                ariaLabel={msgs.get('dropdown.pod.label', locale)}
                label={containers ? containers[0].name : ''}
                onChange={this.handleContainerChange.bind(this)}
                items={containers ? containers.map((container, index) => ({ id: `${container.name}-${index}`, label: container.name, value: container.name})) : []} />
            </div>
          </div>

          {(() => {
            if (!xhrPoll && loading && errors === '') {
              return <Loading withOverlay={false} className='content-spinner' />
            }
            return <ScrollBox className='logs-container__content' content={logs} />
          })()}
        </div>
      </Modal>
    )
  }

  handleContainerChange(data) {
    const containerName = data.selectedItem.label
    const { loading } = this.state
    this.setState({
      xhrPoll: false,
      selectedContainer: containerName,
    })
    if (!loading) {
      const { podName, podNamespace, clusterName } = this.state
      this.fetchLogs(containerName, podName, podNamespace, clusterName)
    }
  }

  reload() {
    const { clusterName, loading, podName, podNamespace, selectedContainer } = this.state
    if (!loading) {
      this.setState({ xhrPoll: true })
      this.fetchLogs(selectedContainer, podName, podNamespace, clusterName)
    }
  }
}

LogsModal.contextTypes = {
  locale: PropTypes.string
}

LogsModal.propTypes = {
  data: PropTypes.object,
  errors: PropTypes.string,
  open: PropTypes.bool,
  type: PropTypes.string
}

export default LogsModal
