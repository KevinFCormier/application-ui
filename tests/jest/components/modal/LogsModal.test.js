/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
"use strict";
jest.mock("../../../../lib/client/apollo-client", () => ({
  getClient: jest.fn(() => {
    return null;
  }),
  getLogs: jest.fn(() => {
    const data = {
      data: {
        logs: [{ name: "aa" }]
      }
    };
    return Promise.resolve(data);
  }),
  getResource: jest.fn(() => {
    const data = {
      data: {
        items: [
          {
            containers: [{ name: "contName" }],
            cluster: {
              metadata: {
                name: "clsName"
              }
            },
            metadata: {
              name: "guestbook-app",
              namespace: "default"
            }
          }
        ]
      }
    };

    return Promise.resolve(data);
  })
}));

import React from "react";
import LogsModal from "../../../../src-web/components/modals/LogsModal";
import { mount } from "enzyme";
import * as reducers from "../../../../src-web/reducers";
import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { resourceModalData, resourceModalLabels } from "./ModalsTestingData";
import toJson from "enzyme-to-json";
import { BrowserRouter } from "react-router-dom";

describe("LogsModal test", () => {
  const handleModalClose = jest.fn();
  const handleModalSubmit = jest.fn();
  const resourceType = { name: "HCMApplication", list: "HCMApplicationList" };
  const preloadedState = window.__PRELOADED_STATE__;
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const middleware = [thunkMiddleware];
  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware))
  );

  it("renders as expected 1", () => {
    const component = mount(
      <BrowserRouter>
        <LogsModal
          data={resourceModalData}
          handleClose={handleModalClose}
          handleSubmit={handleModalSubmit}
          label={resourceModalLabels}
          locale={"en"}
          open={true}
          resourceType={resourceType}
          store={store}
        />
      </BrowserRouter>
    );
    expect(toJson(component.instance())).toMatchSnapshot();
    expect(toJson(component.update())).toMatchSnapshot();
    expect(toJson(component)).toMatchSnapshot();

    component
      .find(".bx--modal")
      .at(0)
      .simulate("click");
    component
      .find(".bx--modal")
      .at(0)
      .simulate("keydown");

    component
      .find(".bx--modal-close")
      .at(0)
      .simulate("click");

    component
      .find(".bx--dropdown")
      .at(0)
      .simulate("click");
    component
      .find(".bx--dropdown")
      .at(0)
      .simulate("keydown");

    component
      .find(".bx--list-box__field")
      .at(0)
      .simulate("click");
    component
      .find(".bx--list-box__field")
      .at(0)
      .simulate("keydown");
  });
});
