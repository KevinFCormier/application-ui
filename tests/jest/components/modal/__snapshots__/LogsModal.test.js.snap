// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogsModal test renders as expected 1 1`] = `undefined`;

exports[`LogsModal test renders as expected 1 2`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <LogsModal
      data={
        Object {
          "__typename": "Application",
          "_uid": "local-cluster/42d4c013-52b9-11ea-bf05-00000a102d26",
          "clusterCount": 0,
          "created": "2019-12-11T01:24:02Z",
          "dashboard": null,
          "hubSubscriptions": Array [],
          "kind": "HCMApplication",
          "metadata": Object {
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/sample/applications/samplebook-gbap",
          },
          "name": "samplebook-gbap",
          "namespace": "sample",
          "podStatusCount": Object {},
          "remoteSubscriptionStatusCount": Object {},
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/sample/applications/samplebook-gbap",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "pprimaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <Modal
        aria-label="logs"
        className="logs"
        iconDescription="close the modal"
        id="view-logs-modal"
        modalHeading=""
        modalLabel="Pod"
        onKeyDown={[Function]}
        onRequestClose={[Function]}
        onRequestSubmit={[Function]}
        open={true}
        passiveModal={true}
        primaryButtonDisabled={true}
        role="region"
        selectorPrimaryFocus="[data-modal-primary-focus]"
      >
        <div
          aria-label="logs"
          className="bx--modal is-visible logs"
          id="view-logs-modal"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          onTransitionEnd={[Function]}
          role="presentation"
          tabIndex={-1}
        >
          <div
            className="bx--modal-container"
            role="dialog"
          >
            <div
              className="bx--modal-header"
            >
              <button
                className="bx--modal-close"
                onClick={[Function]}
                type="button"
              >
                <Icon
                  className="bx--modal-close__icon"
                  description="close the modal"
                  fillRule="evenodd"
                  icon={
                    Object {
                      "height": "10",
                      "id": "icon--close",
                      "name": "icon--close",
                      "styles": "",
                      "svgData": Object {
                        "circles": "",
                        "ellipses": "",
                        "paths": Array [
                          Object {
                            "d": "M6.32 5L10 8.68 8.68 10 5 6.32 1.32 10 0 8.68 3.68 5 0 1.32 1.32 0 5 3.68 8.68 0 10 1.32 6.32 5z",
                            "fill-rule": "nonzero",
                          },
                        ],
                        "polygons": "",
                        "polylines": "",
                        "rects": "",
                      },
                      "tags": "icon--close",
                      "viewBox": "0 0 10 10",
                      "width": "10",
                    }
                  }
                  role="img"
                >
                  <svg
                    alt="close the modal"
                    aria-label="close the modal"
                    className="bx--modal-close__icon"
                    fillRule="evenodd"
                    height="10"
                    role="img"
                    viewBox="0 0 10 10"
                    width="10"
                  >
                    <title>
                      close the modal
                    </title>
                    <path
                      d="M6.32 5L10 8.68 8.68 10 5 6.32 1.32 10 0 8.68 3.68 5 0 1.32 1.32 0 5 3.68 8.68 0 10 1.32 6.32 5z"
                      key="key0"
                    />
                  </svg>
                </Icon>
              </button>
              <h4
                className="bx--modal-header__label"
              >
                Pod
              </h4>
              <h2
                className="bx--modal-header__heading"
              />
            </div>
            <div
              className="bx--modal-content"
            >
              <div
                className="logs-container"
              >
                <div
                  className="logs-container__actions"
                >
                  <div
                    className="dropdown-container"
                  >
                    <DropdownV2
                      disabled={false}
                      itemToElement={null}
                      itemToString={[Function]}
                      items={Array []}
                      label=""
                      light={false}
                      onChange={[Function]}
                      type="default"
                    >
                      <Downshift
                        breakingChanges={Object {}}
                        defaultHighlightedIndex={null}
                        defaultInputValue=""
                        defaultIsOpen={false}
                        defaultSelectedItem={null}
                        environment={[Window]}
                        getA11yStatusMessage={[Function]}
                        itemToString={[Function]}
                        onChange={[Function]}
                        onInputValueChange={[Function]}
                        onOuterClick={[Function]}
                        onSelect={[Function]}
                        onStateChange={[Function]}
                        onUserAction={[Function]}
                        selectedItemChanged={[Function]}
                        stateReducer={[Function]}
                      >
                        <ListBox
                          ariaLabel="Choose an item"
                          className="bx--dropdown"
                          disabled={false}
                          innerRef={[Function]}
                          type="default"
                        >
                          <div
                            aria-label="Choose an item"
                            className="bx--dropdown bx--list-box"
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            role="listbox"
                            tabIndex="0"
                          >
                            <ListBoxField
                              aria-expanded={false}
                              aria-haspopup={true}
                              aria-label="open menu"
                              data-toggle={true}
                              disabled={false}
                              onBlur={[Function]}
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="button"
                              type="button"
                            >
                              <div
                                aria-expanded={false}
                                aria-haspopup={true}
                                aria-label="open menu"
                                className="bx--list-box__field"
                                data-toggle={true}
                                disabled={false}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onKeyDown={[Function]}
                                role="button"
                                tabIndex="0"
                                type="button"
                              >
                                <span
                                  className="bx--list-box__label"
                                  htmlFor="downshift-0-input"
                                />
                                <ListBoxMenuIcon
                                  isOpen={false}
                                  translateWithId={[Function]}
                                >
                                  <div
                                    className="bx--list-box__menu-icon"
                                  >
                                    <Icon
                                      alt="Open menu"
                                      description="Open menu"
                                      fillRule="evenodd"
                                      icon={
                                        Object {
                                          "height": "5",
                                          "id": "icon--caret--down",
                                          "name": "icon--caret--down",
                                          "styles": "",
                                          "svgData": Object {
                                            "circles": "",
                                            "ellipses": "",
                                            "paths": Array [
                                              Object {
                                                "d": "M0 0l5 4.998L10 0z",
                                              },
                                            ],
                                            "polygons": "",
                                            "polylines": "",
                                            "rects": "",
                                          },
                                          "tags": "icon--caret--down",
                                          "viewBox": "0 0 10 5",
                                          "width": "10",
                                        }
                                      }
                                      role="img"
                                    >
                                      <svg
                                        alt="Open menu"
                                        aria-label="Open menu"
                                        fillRule="evenodd"
                                        height="5"
                                        role="img"
                                        viewBox="0 0 10 5"
                                        width="10"
                                      >
                                        <title>
                                          Open menu
                                        </title>
                                        <path
                                          d="M0 0l5 4.998L10 0z"
                                          key="key0"
                                        />
                                      </svg>
                                    </Icon>
                                  </div>
                                </ListBoxMenuIcon>
                              </div>
                            </ListBoxField>
                          </div>
                        </ListBox>
                      </Downshift>
                    </DropdownV2>
                  </div>
                </div>
                <Loading
                  active={true}
                  className="content-spinner"
                  small={false}
                  withOverlay={false}
                >
                  <div
                    aria-live="assertive"
                    className="bx--loading content-spinner"
                  >
                    <svg
                      className="bx--loading__svg"
                      viewBox="-75 -75 150 150"
                    >
                      <circle
                        cx="0"
                        cy="0"
                        r="37.5"
                      />
                    </svg>
                  </div>
                </Loading>
              </div>
            </div>
          </div>
        </div>
      </Modal>
    </LogsModal>
  </Router>
</BrowserRouter>
`;

exports[`LogsModal test renders as expected 1 3`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <LogsModal
      data={
        Object {
          "__typename": "Application",
          "_uid": "local-cluster/42d4c013-52b9-11ea-bf05-00000a102d26",
          "clusterCount": 0,
          "created": "2019-12-11T01:24:02Z",
          "dashboard": null,
          "hubSubscriptions": Array [],
          "kind": "HCMApplication",
          "metadata": Object {
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/sample/applications/samplebook-gbap",
          },
          "name": "samplebook-gbap",
          "namespace": "sample",
          "podStatusCount": Object {},
          "remoteSubscriptionStatusCount": Object {},
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/sample/applications/samplebook-gbap",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "pprimaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <Modal
        aria-label="logs"
        className="logs"
        iconDescription="close the modal"
        id="view-logs-modal"
        modalHeading=""
        modalLabel="Pod"
        onKeyDown={[Function]}
        onRequestClose={[Function]}
        onRequestSubmit={[Function]}
        open={true}
        passiveModal={true}
        primaryButtonDisabled={true}
        role="region"
        selectorPrimaryFocus="[data-modal-primary-focus]"
      >
        <div
          aria-label="logs"
          className="bx--modal is-visible logs"
          id="view-logs-modal"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          onTransitionEnd={[Function]}
          role="presentation"
          tabIndex={-1}
        >
          <div
            className="bx--modal-container"
            role="dialog"
          >
            <div
              className="bx--modal-header"
            >
              <button
                className="bx--modal-close"
                onClick={[Function]}
                type="button"
              >
                <Icon
                  className="bx--modal-close__icon"
                  description="close the modal"
                  fillRule="evenodd"
                  icon={
                    Object {
                      "height": "10",
                      "id": "icon--close",
                      "name": "icon--close",
                      "styles": "",
                      "svgData": Object {
                        "circles": "",
                        "ellipses": "",
                        "paths": Array [
                          Object {
                            "d": "M6.32 5L10 8.68 8.68 10 5 6.32 1.32 10 0 8.68 3.68 5 0 1.32 1.32 0 5 3.68 8.68 0 10 1.32 6.32 5z",
                            "fill-rule": "nonzero",
                          },
                        ],
                        "polygons": "",
                        "polylines": "",
                        "rects": "",
                      },
                      "tags": "icon--close",
                      "viewBox": "0 0 10 10",
                      "width": "10",
                    }
                  }
                  role="img"
                >
                  <svg
                    alt="close the modal"
                    aria-label="close the modal"
                    className="bx--modal-close__icon"
                    fillRule="evenodd"
                    height="10"
                    role="img"
                    viewBox="0 0 10 10"
                    width="10"
                  >
                    <title>
                      close the modal
                    </title>
                    <path
                      d="M6.32 5L10 8.68 8.68 10 5 6.32 1.32 10 0 8.68 3.68 5 0 1.32 1.32 0 5 3.68 8.68 0 10 1.32 6.32 5z"
                      key="key0"
                    />
                  </svg>
                </Icon>
              </button>
              <h4
                className="bx--modal-header__label"
              >
                Pod
              </h4>
              <h2
                className="bx--modal-header__heading"
              />
            </div>
            <div
              className="bx--modal-content"
            >
              <div
                className="logs-container"
              >
                <div
                  className="logs-container__actions"
                >
                  <div
                    className="dropdown-container"
                  >
                    <DropdownV2
                      disabled={false}
                      itemToElement={null}
                      itemToString={[Function]}
                      items={Array []}
                      label=""
                      light={false}
                      onChange={[Function]}
                      type="default"
                    >
                      <Downshift
                        breakingChanges={Object {}}
                        defaultHighlightedIndex={null}
                        defaultInputValue=""
                        defaultIsOpen={false}
                        defaultSelectedItem={null}
                        environment={[Window]}
                        getA11yStatusMessage={[Function]}
                        itemToString={[Function]}
                        onChange={[Function]}
                        onInputValueChange={[Function]}
                        onOuterClick={[Function]}
                        onSelect={[Function]}
                        onStateChange={[Function]}
                        onUserAction={[Function]}
                        selectedItemChanged={[Function]}
                        stateReducer={[Function]}
                      >
                        <ListBox
                          ariaLabel="Choose an item"
                          className="bx--dropdown"
                          disabled={false}
                          innerRef={[Function]}
                          type="default"
                        >
                          <div
                            aria-label="Choose an item"
                            className="bx--dropdown bx--list-box"
                            onClick={[Function]}
                            onKeyDown={[Function]}
                            role="listbox"
                            tabIndex="0"
                          >
                            <ListBoxField
                              aria-expanded={false}
                              aria-haspopup={true}
                              aria-label="open menu"
                              data-toggle={true}
                              disabled={false}
                              onBlur={[Function]}
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="button"
                              type="button"
                            >
                              <div
                                aria-expanded={false}
                                aria-haspopup={true}
                                aria-label="open menu"
                                className="bx--list-box__field"
                                data-toggle={true}
                                disabled={false}
                                onBlur={[Function]}
                                onClick={[Function]}
                                onKeyDown={[Function]}
                                role="button"
                                tabIndex="0"
                                type="button"
                              >
                                <span
                                  className="bx--list-box__label"
                                  htmlFor="downshift-0-input"
                                />
                                <ListBoxMenuIcon
                                  isOpen={false}
                                  translateWithId={[Function]}
                                >
                                  <div
                                    className="bx--list-box__menu-icon"
                                  >
                                    <Icon
                                      alt="Open menu"
                                      description="Open menu"
                                      fillRule="evenodd"
                                      icon={
                                        Object {
                                          "height": "5",
                                          "id": "icon--caret--down",
                                          "name": "icon--caret--down",
                                          "styles": "",
                                          "svgData": Object {
                                            "circles": "",
                                            "ellipses": "",
                                            "paths": Array [
                                              Object {
                                                "d": "M0 0l5 4.998L10 0z",
                                              },
                                            ],
                                            "polygons": "",
                                            "polylines": "",
                                            "rects": "",
                                          },
                                          "tags": "icon--caret--down",
                                          "viewBox": "0 0 10 5",
                                          "width": "10",
                                        }
                                      }
                                      role="img"
                                    >
                                      <svg
                                        alt="Open menu"
                                        aria-label="Open menu"
                                        fillRule="evenodd"
                                        height="5"
                                        role="img"
                                        viewBox="0 0 10 5"
                                        width="10"
                                      >
                                        <title>
                                          Open menu
                                        </title>
                                        <path
                                          d="M0 0l5 4.998L10 0z"
                                          key="key0"
                                        />
                                      </svg>
                                    </Icon>
                                  </div>
                                </ListBoxMenuIcon>
                              </div>
                            </ListBoxField>
                          </div>
                        </ListBox>
                      </Downshift>
                    </DropdownV2>
                  </div>
                </div>
                <Loading
                  active={true}
                  className="content-spinner"
                  small={false}
                  withOverlay={false}
                >
                  <div
                    aria-live="assertive"
                    className="bx--loading content-spinner"
                  >
                    <svg
                      className="bx--loading__svg"
                      viewBox="-75 -75 150 150"
                    >
                      <circle
                        cx="0"
                        cy="0"
                        r="37.5"
                      />
                    </svg>
                  </div>
                </Loading>
              </div>
            </div>
          </div>
        </div>
      </Modal>
    </LogsModal>
  </Router>
</BrowserRouter>
`;
